cmd/podman/restore.go

```diff
@@ -46,6 +56,94 @@ var (
 	}
 )
 
+func importFromJson(filePath string) ([]byte, error) {
+	jsonFile, err := os.Open(filePath)
+	if err != nil {
+		return nil, err
+	}
+	defer jsonFile.Close()
+
+	return ioutil.ReadAll(jsonFile)
+}
+
+func importCheckpoint(c *cli.Context, runtime *libpod.Runtime, input string) ([]*libpod.Container, error) {
+	// First get the container definition from the
+	// tarball to a temporary directory
+	archiveFile, err := os.Open(input)
+	if err != nil {
+		return nil, err
+	}
+	defer archiveFile.Close()
+	options := &archive.TarOptions{
+		// Here we only need the files config.dump and spec.dump
+		ExcludePatterns: []string{
+			"checkpoint",
+			"artifacts",
+			"ctr.log",
+			"network.status",
+		},
+	}
+	dir, err := ioutil.TempDir("", "checkpoint")
+	if err != nil {
+		return nil, err
+	}
+	defer os.RemoveAll(dir)
+	err = archive.Untar(archiveFile, dir, options)
+	if err != nil {
+		return nil, err
+	}
+
+	// Load spec.dump from temporary directory
+	content, err := importFromJson(filepath.Join(dir, "spec.dump"))
+	if err != nil {
+		return nil, err
+	}
+	spec := new(spec.Spec)
+	err = json.Unmarshal([]byte(content), spec)
+	if err != nil {
+		return nil, err
+	}
+
+	// Load config.dump from temporary directory
+	content, err = importFromJson(filepath.Join(dir, "config.dump"))
+	if err != nil {
+		return nil, err
+	}
+	config := new(libpod.ContainerConfig)
+	err = json.Unmarshal([]byte(content), config)
+	if err != nil {
+		return nil, err
+	}
+
```

Comment:
The code that reads the content of `config.dump` and `spec.dump` looks very similar. Would it be possible to add a second argument to the `importFromJson` function and move the call to `json.Unmarshal` into that function?

Reply:
I tried but I failed. Not sure how I can pass the type of the data that should be unmarshalled into a function.

Reply:
What about something like this:
```diff
diff --git a/cmd/podman/restore.go b/cmd/podman/restore.go
index 5e1045d1..598fd137 100644
--- a/cmd/podman/restore.go
+++ b/cmd/podman/restore.go
@@ -56,14 +56,20 @@ var (
 	}
 )
 
-func importFromJson(filePath string) ([]byte, error) {
+func importFromJSON(filePath string, v interface{}) error {
 	jsonFile, err := os.Open(filePath)
 	if err != nil {
-		return nil, err
+		return err
 	}
 	defer jsonFile.Close()
 
-	return ioutil.ReadAll(jsonFile)
+	content, err := ioutil.ReadAll(jsonFile)
+	err = json.Unmarshal([]byte(content), v)
+	if err != nil {
+		return err
+	}
+
+	return nil
 }
 
 func importCheckpoint(c *cli.Context, runtime *libpod.Runtime, input string) ([]*libpod.Container, error) {
@@ -94,24 +100,14 @@ func importCheckpoint(c *cli.Context, runtime *libpod.Runtime, input string) ([]
 	}
 
 	// Load spec.dump from temporary directory
-	content, err := importFromJson(filepath.Join(dir, "spec.dump"))
-	if err != nil {
-		return nil, err
-	}
 	spec := new(spec.Spec)
-	err = json.Unmarshal([]byte(content), spec)
-	if err != nil {
+	if err := importFromJSON(filepath.Join(dir, "spec.dump"), spec); err != nil {
 		return nil, err
 	}
 
 	// Load config.dump from temporary directory
-	content, err = importFromJson(filepath.Join(dir, "config.dump"))
-	if err != nil {
-		return nil, err
-	}
 	config := new(libpod.ContainerConfig)
-	err = json.Unmarshal([]byte(content), config)
-	if err != nil {
+	if err = importFromJSON(filepath.Join(dir, "config.dump"), config); err != nil {
 		return nil, err
 	}
```

Reply:
Perfect, thanks, that works. I was trying to use the interface as a return value and that did not work for me. This way it works. Do you want to send me a patch for this series or should I just add it to my patch?

Reply:
> Do you want to send me a patch for this series or should I just add it to my patch?

You can just add it to your patch, if that is OK with you.